controllers/azurecluster_controller_test.go

```diff
@@ -133,3 +280,164 @@ func TestAzureClusterReconcilePaused(t *testing.T) {
+			createAzureClusterService: getFakeAzureClusterServiceWithTransientError,
+			cache:                     &scope.ClusterCache{},
+			expectedResult:            reconcile.Result{},
+		},
+		"should fail to delete for non-transient errors": {
+			createAzureClusterService: getFakeAzureClusterServiceWithGeneralError,
+			cache:                     &scope.ClusterCache{},
+			expectedErr:               "error deleting AzureCluster",
+		},
+	}
+
+	for name, c := range cases {
+		tc := c
+		t.Run(name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			reconciler, clusterScope, err := getClusterReconcileInputs(tc)
+			g.Expect(err).NotTo(HaveOccurred())
+
+			result, err := reconciler.reconcileDelete(context.Background(), clusterScope)
+			g.Expect(result).To(Equal(tc.expectedResult))
+
+			if tc.expectedErr != "" {
+				g.Expect(err).To(HaveOccurred())
+				g.Expect(err.Error()).To(ContainSubstring(tc.expectedErr))
+			} else {
+				g.Expect(err).NotTo(HaveOccurred())
+			}
+		})
+	}
+}
+
+func getFakeAzureClusterService(clusterScope *scope.ClusterScope) (*azureClusterService, error) {
+	cache := resourceskus.NewStaticCache([]armcompute.ResourceSKU{}, clusterScope.Location())
+
+	return getDefaultAzureClusterService(clusterScope, cache), nil
+}
+
+func getFakeAzureClusterServiceWithFailure(_ *scope.ClusterScope) (*azureClusterService, error) {
+	return nil, errors.New("failed to create AzureClusterService")
+}
+
+func getFakeAzureClusterServiceWithTerminalError(clusterScope *scope.ClusterScope) (*azureClusterService, error) {
+	cache := resourceskus.NewStaticCache([]armcompute.ResourceSKU{}, clusterScope.Location())
+
+	acs := getDefaultAzureClusterService(clusterScope, cache)
+	acs.Reconcile = func(context.Context) error {
+		return azure.WithTerminalError(errors.New("failed to reconcile AzureCluster"))
+	}
+
+	return acs, nil
+}
+
+func getFakeAzureClusterServiceWithTransientError(clusterScope *scope.ClusterScope) (*azureClusterService, error) {
+	cache := resourceskus.NewStaticCache([]armcompute.ResourceSKU{}, clusterScope.Location())
+
+	acs := getDefaultAzureClusterService(clusterScope, cache)
+	acs.Reconcile = func(context.Context) error {
+		return azure.WithTransientError(errors.New("failed to reconcile AzureCluster"), 10*time.Second)
+	}
+
+	return acs, nil
+}
+
+func getFakeAzureClusterServiceWithGeneralError(clusterScope *scope.ClusterScope) (*azureClusterService, error) {
+	cache := resourceskus.NewStaticCache([]armcompute.ResourceSKU{}, clusterScope.Location())
+
+	acs := getDefaultAzureClusterService(clusterScope, cache)
+	acs.Reconcile = func(context.Context) error {
+		return errors.New("foo error")
+	}
+	acs.Pause = func(context.Context) error {
+		return errors.New("foo error")
+	}
+	acs.Delete = func(context.Context) error {
+		return errors.New("foo error")
+	}
+
+	return acs, nil
+}
+
+func getDefaultAzureClusterService(clusterScope *scope.ClusterScope, cache *resourceskus.Cache) *azureClusterService {
```

Comment:
Instead of hard-coding the parameters here, I think it would be cleaner to use the changes pattern so you can change any of the fields if needed. This would also eliminate the need for the other wrapper functions you made like `getFakeAzureClusterServiceWithTerminalError`. An example of how this is done is here:

- Function: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/controllers/asosecret_controller_test.go#L320
- Using the function: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/controllers/asosecret_controller_test.go#L70